ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Mempool/Query.hs

```diff
@@ -1,29 +1,89 @@
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE FlexibleContexts #-}
 
 -- | Queries to the mempool
-module Ouroboros.Consensus.Mempool.Query (pureGetSnapshotFor) where
+module Ouroboros.Consensus.Mempool.Query (
+    implGetSnapshotFor
+  , pureGetSnapshotFor
+  ) where
 
-import           Ouroboros.Consensus.HeaderValidation
+#if __GLASGOW_HASKELL__ < 910
+import           Data.Foldable (foldl')
+#endif
+import           Ouroboros.Consensus.Block.Abstract
 import           Ouroboros.Consensus.Ledger.Abstract
 import           Ouroboros.Consensus.Ledger.SupportsMempool
+import           Ouroboros.Consensus.Ledger.Tables.Utils (emptyLedgerTables)
 import           Ouroboros.Consensus.Mempool.API
 import           Ouroboros.Consensus.Mempool.Capacity
 import           Ouroboros.Consensus.Mempool.Impl.Common
+import qualified Ouroboros.Consensus.Mempool.TxSeq as TxSeq
+import           Ouroboros.Consensus.Util.IOLike
+
+implGetSnapshotFor ::
+     ( IOLike m
+     , LedgerSupportsMempool blk
+     , HasTxId (GenTx blk)
+     )
+  => MempoolEnv m blk
+  -> SlotNo -- ^ Get snapshot for this slot number (usually the current slot)
+  -> TickedLedgerState blk DiffMK -- ^ The ledger state at 'pt' ticked to 'slot'
+  -> (LedgerTables (LedgerState blk) KeysMK -> m (LedgerTables (LedgerState blk) ValuesMK))
+      -- ^ A function that returns values corresponding to the given keys for
+      -- the unticked ledger state at 'pt'.
+  -> m (MempoolSnapshot blk)
+implGetSnapshotFor mpEnv slot ticked readUntickedTables = do
+  is <- atomically $ readTMVar istate
+  if pointHash (isTip is) == castHash (getTipHash ticked) &&
+     isSlotNo is == slot
+    then
+      -- We are looking for a snapshot exactly for the ledger state we already
+      -- have cached, then just return it.
+      pure . snapshotFromIS $ is
+    else do
+       let keys = foldl' (<>) emptyLedgerTables
+                $ map getTransactionKeySets
+                $ [ txForgetValidated . TxSeq.txTicketTx $ tx
+                  | tx <- TxSeq.toList $ isTxs is
+                  ]
+       values <- readUntickedTables keys
```

Comment:
I see, you're proposing some sort of cache for ledger tables. I like the idea, though we should investigate whether we can come up with a sensible design. It might also make more sense if that functionality lives in the LedgerDB, and not in the mempool, similarly to how the block cache is in the ImmutableDB. Let's create an issue for this cache feature and look into it at some point after merging this PR

Reply:
#1303